README
					README 
					------
			
					expirer
					=======
					version.0.1
	
*About expirer
*How to install
*User Guide and Documents


About expirer:
==============
expirer is a file expiry tool. It allows admin user to set expiry time for specific files and 
automagically removes them when it expires.

How to install:
============== 
see INSTALL file

How expirer works?
-----------------

When user sets file name and its time using 'expirer',time value recorded in file inode's dtime field.
Then the values filename,time,md5sum stored on  Berkeley DB (/etc/expirer/info.db) with time as key.
And restart the daemon process 'expirerd'.

expirerd is a daemon process which keeps reading DB file and when it finds entry which matches current
time, it compares the disk file owner,inode,md5sum with DB values. If there is a perfect match daemon
process unlinks the filepath and fetches the next DB record and repeats the process.

Why not use a simple scrit with cron ? 
---------------------------------------

Yes. You can achieve same effect with a simple bash/python scripts. One possible advantage(?) over 
such script is:

-    With bash/python scripts you need to store the file details at certain db. If there are 50 people 
     and each of them sets 1 file location and time to the db. Now db goes corrupt or missing. You need 
     to re-enter the value manually 50 times. With 'expirer' you can perform a simple rescan to re-create 
     the db entires,no need to manually set the file expiry time again!

Why you need such tool ?
-----------------------

- We have project where we(root) share files with users in a common directory. Just needed to ensure,that
  these files are deleted from common share once the user copies them to their home directories.

Why inode's dtime is used instead of acl?
----------------------------------------

acl can be used, If I'm not wrong sometimes(or always?) setting 'acl' entries consume 1 block. 
using dtime this can be prevented.

If you find some other field (may be i_ctime_extra ? ) which will be more useful or right place
holder than dtime. Please let me know. Thanks in advance!

Do you have an expirer GUI?
---------------------------

Glad you asked for it :D Yes!!!. We do have one based on Python-Kivy. Simply run 'expirer-gui' to use it.

How to use the tool
===================
 	
How to set expiry time for files?
---------------------------------

You need to pass devicename,absolute filepath and minutes as arguments to expirer binary.

Usage: expirer   [-l list] [-d devicename -f filepath -t minutes] [-s -d devicename -m mountpoint] [-c cancel -f filepath]

<code>
# expirer -d /dev/sda7 -f /home/laks/file.exp6 -t 10

 File /home/laks/file.exp6 will expire in 10 minutes : Sat Dec 21 12:38:17 2013
</code>

For example, above command ensure the file "/home/laks/file.exp6" expires in 10 minutes
which happens to be "Sat Dec 21 12:38:17 2013".

How to view list of expiry files and it details?
-----------------------------------------------
<code>
# expirer -l

	Filename	      Expires on

	--------	      ----------

	/home/laks/file.exp6	Sat Dec 21 12:38:17 2013	

	/home/laks/file.exp10	Sat Dec 21 19:15:39 2013
</code>
How to Cancel an entry from the expiry job list?
------------------------------------------------
Suppose, your mind changed, now you won't want the file (file.exp10) to get expired. In order to cancel the previously
applied expiry settings run the following command:
<code>
# expirer -c -f /home/laks/file.exp10
</code>
How to scan a device for expiry files ?
--------------------------------------

If you want to recreate the database use -s option rescan the partition.
<code>
# expirer -s -d devicename -m mounteddir
</code>

example:
<code>
#expirer -s -d /dev/sda7 -m /home
</code>

Will scan the drive /dev/sda7 which was mounted on /home directory.

Note: Run this command only when /etc/expirer/info.db is corrupt or accidentally 
removed.


User Guide and Documents
========================
For more screencasting/manuals,checkout  www.giis.co.in

You can get me at  <lakshmipathi.g@giis.co.in>

-Lakshmipathi.G
www.giis.co.in
-Dec 27,2013.
==================================EOF==================================


ChangeLog
................................................................................................,

2013-12-27  Lakshmipathi.G  <lakshmipathi.g@giis.co.in>

	* Release of expirer v0.1
	
			- Uses ext2fs library and bdb.


test/test.sh
for i in {1..10};do
dd if=/dev/urandom of=/home/laks/testdir/file$i.txt bs=1024 count=1
sync
sync
sync
sync
sleep 5
/usr/sbin/expirer -d  /dev/sda7 -f /home/laks/testdir/file$i.txt -t $i

done


src/expirerd.c
/*
* /expirer/expirerd.c - At specific time period, automagically the files will expire.
*
* Copyright (C) 2013 Lakshmipathi.G <lakshmipathi.g@giis.co.in>
* Visit www.giis.co.in for manuals or docs.
*/

#include <db.h>
#include <sys/time.h>
#include <string.h>
#include <stdlib.h>
#include <sys/stat.h>
#include <signal.h>
#define DB_PATH "/etc/expirer/info.db"


struct expirer_data {
	unsigned long int inode_number,expiry_time;
	int owner;
	char *file_path;
	char *sha256sum;
};

void restart_process(char **args){
	int childpid;

	childpid = fork ();
	if (childpid < 0) {
		perror ("fork failed");
	} else if (childpid  == 0) {
		printf ("new process %d", getpid());
		int rv = execve (args[0], args, NULL);
		if (rv == -1) {
			perror ("execve");
			exit (EXIT_FAILURE);
		}

	} else {
		sleep (5);
		printf ("killing %d\n", getpid());
		kill (getpid (), SIGTERM);
	}
}

char md5sum[34];
volatile sig_atomic_t signal_received = 0;
void sighandler (int signum) {
	signal_received = 1;
}


void compute_md5sum(char *filelocation){
	FILE *pf;
	char md5_cmd[512];
	extern char md5sum[34];
	//Recompute md5  of recovered file
	memset(md5_cmd,'\0',512);
	memset(md5sum,'\0',34);
	sprintf(md5_cmd,"md5sum %s",filelocation);
	pf=popen(md5_cmd,"r");
	if(!pf){
		fprintf(stderr,"Could not open pipe");
		return ;
	}

	//get data
	fgets(md5sum, 34 , pf);

	if (pclose(pf) != 0)
		fprintf(stderr," Error: close Failed.");
}

time_t expirer_current_time();
validate_entries(struct expirer_data user){
	struct stat filestat;
	extern char md5sum[34];

	if (stat(user.file_path, &filestat) == -1){
		printf("file %s not found : disk-file already renamed/unlinked.\n",user.file_path);
		return 1;
	}
	else{
		if (user.inode_number != filestat.st_ino || user.owner != filestat.st_uid){
			printf("Either file inode is re-used %lu or\
					Owner has been changed %d.\
					So Skipping this file%d\n",user.inode_number,user.owner,user.file_path);
			return 1;
		}
		compute_md5sum(user.file_path);
		if((strcmp (user.sha256sum,md5sum)))
			return 1; //checksum mismatch

	}

	return 0;
}
// get current time in seconds.
time_t expirer_current_time(){
	struct timeval tv;
	struct timezone tz;

	gettimeofday(&tv, &tz);
	return tv.tv_sec;
}

int main(int argc,char* argv[]){

	unsigned long int etime;
	DBT key, data;
	DB *my_database;
	struct expirer_data user;
	char *buffer,*key_buf;
	int ret;
	u_int32_t flags;
	struct stat filestat;
	unsigned long int cur_time=0;
	FILE *fd;
	char buf[32];
	DBC *cursorp;

	if (signal (SIGUSR1, sighandler) == SIG_ERR) {
		perror ("signal failed");
	}
forever_loop:
	if (signal_received) {
		goto close_db;
	}
	// get pid and log it
	fd = fopen("/var/run/expirerd.pid","w");
	sprintf(buf,"%d",getpid());
	fwrite(buf,sizeof(int),1,fd);
	fclose(fd);

	if (stat(DB_PATH, &filestat) == -1){
		printf("file not found");
		return 1;
	}
	if (filestat.st_uid !=0 || filestat.st_mode != 33184 ){
		printf("Wrong DB permissions uid=%d mode=%d",filestat.st_uid,filestat.st_mode);
		return 1;
	}

	ret = db_create(&my_database, NULL, 0);
	if (ret != 0) {
		printf("\n db_create failed.");
		return 1;
	}
	flags = DB_CREATE;   

	ret = my_database->open(my_database,NULL,DB_PATH,NULL,DB_BTREE,flags,0);
	if (ret != 0) {
		printf("\n db_open failed.");
		return 1;
	}


	/* Zero out the DBTs before using them. */
	memset(&key, 0, sizeof(DBT));
	memset(&data, 0, sizeof(DBT));

	/* Get a cursor */
	my_database->cursor(my_database,NULL,&cursorp,0);

	ret = cursorp->get(cursorp,&key,&data,DB_NEXT);

	if (ret < 0) {
		printf("\n Empty Db");
		sleep(120);//ideally sleep until USR1 received!
	}else{
		key_buf = key.data;
		buffer = data.data;
		etime = *((unsigned long *)key.data);

		user.owner = *((int *)data.data);
		user.inode_number = *((unsigned long *)(buffer+sizeof(int)));
		user.expiry_time = *((unsigned long*)(buffer + sizeof(int) +sizeof(unsigned long int)));
		user.file_path = buffer + (sizeof(unsigned long int) * 2) + sizeof(int); 
		user.sha256sum = buffer + (sizeof(unsigned long int) * 2) + sizeof(int) + strlen(user.file_path) +1; 


		cur_time = expirer_current_time();
		while (user.expiry_time > cur_time){
			sleep(10);
			cur_time = expirer_current_time();
			if (signal_received) 
				goto close_db;
		}

		//validate current disk file entries match db stat
		if (validate_entries(user) == 0){
			unlink(user.file_path);
			cursorp->del(cursorp, 0);
		}else //Ignore- perm changed or file renamed/deleted - checksum mismatch.
		{
			cursorp->del(cursorp, 0);
		}

	}
	/* Cursors must be closed */
close_db:
	if (cursorp != NULL)
		cursorp->close(cursorp); 

	if (my_database != NULL)
		my_database->close(my_database, 0);
	if(signal_received){
		signal_received = 0;
		restart_process (argv);
	}

	goto forever_loop;
}


src/expirer.c
/*
* /expirer/expirer.c - At specific time period, automagically the files will expire.
*
* Copyright (C) 2013 Lakshmipathi.G <lakshmipathi.g@giis.co.in>
* Visit www.giis.co.in for manuals or docs.
*/

#define _GNU_SOURCE
#include <dirent.h>     /* Defines DT_* constants */
#include <fcntl.h>
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>
#include <sys/syscall.h>
#include <sys/time.h>
#include <sys/time.h>
#include <ext2fs/ext2fs.h>
#include <ext2fs/ext2_io.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <pwd.h>
#include <errno.h>
#include <assert.h>    /* assert */
#include <libgen.h>
/* argp */
#include <argp.h>
// bdb routines
#include <db.h>
#include <string.h>
#include <stdlib.h>
#include <signal.h>


#define handle_error(msg) \
do { perror(msg); exit(EXIT_FAILURE); } while (0)
#define BUF_SIZE 1024
#define TRUE 1
#define FALSE 0
#define MAXPATHLEN 4096
#define DB_PATH "/etc/expirer/info.db"



struct arguments
{
	char *device_name,*filepath,*mounteddir;
	char *deltime;		    /* 1-device 2-filepath 3-time 4-scan*/
	int flag;
};
struct ext2_struct_inode_scan {
	errcode_t		magic;
	ext2_filsys		fs;
	ext2_ino_t		current_inode;
	blk64_t			current_block;
	dgrp_t			current_group;
	ext2_ino_t		inodes_left;
	blk_t			blocks_left;
	dgrp_t			groups_left;
	blk_t			inode_buffer_blocks;
	char *			inode_buffer;
	int			inode_size;
	char *			ptr;
	int			bytes_left;
	char			*temp_buffer;
	errcode_t		(*done_group)(ext2_filsys fs,
					      ext2_inode_scan scan,
					      dgrp_t group,
					      void * priv_data);
	void *			done_group_data;
	int			bad_block_ptr;
	int			scan_flags;
	int			reserved[6];
};

struct node {
	unsigned long file_inode,parent_inode,expiry_time;
	char *pathname;
	struct node* next;
	};
struct node *head=NULL;

static struct argp_option options[] =
{  
	{"device",   'd', "Device name", 0,"Partition where the file resides."},
	{"file",   'f', "Absoulte path", 0,"Absoulte path of expiry file."},
	{"scan", 's', 0, 0, "Re-create the database."},
	{"mounteddir",   'm', "Mount point", 0,"Directory where partition mounted."},
	{"time", 't', "Minutes", 0, "Remaining minutes before the file expires."},
	{"list", 'l', 0, 0, "List file  details."},
	{"cancel", 'c', 0, 0, "Cancel an entry."},
	{0}
};
int EXT2_BLOCK_SIZE;
char device[75];
const char *argp_program_version = "expirer 0.1 (27-12-2012) ";
const char *argp_program_bug_address = "<lakshmipathi.g@giis.co.in>";


/* Functions involved. */
static error_t parse_opt (int, char *, struct argp_state *); 
time_t expirer_current_time();
void expirer_read_inode(ext2_filsys,struct stat * ,struct ext2_inode *);
void expirer_write_inode(ext2_filsys,struct stat * ,struct ext2_inode *);
void expirer_set_dtime(struct ext2_inode *, int);
void expirer_computer_md5sum(char *);
int expirer_list_files(int,char *);
void expirer_file_inode_list_push(struct node**,unsigned long,unsigned long ,unsigned long );
char* expirer_get_abs_path(char *);
void expirer_add_list_to_bdb(struct node*);
void expirer_compute_md5sum(char *);

static char args_doc[] = "";
static char doc[] = "\n\nUsage: expirer [-s scan -d devicename -m mounteddir] [-l list] [-c -f filepath] [-d devicename -f filepath -t minutes]\n\nexpirer - File expiry tool for ext{2,3,4} file system.(http://www.giis.co.in)\n\n";
static struct argp argp = {options, parse_opt, args_doc, doc};


static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	struct arguments *arguments = state->input;

	switch (key)
	{
		case 'd':
			arguments->device_name=arg;
			break;
		case 'f':
			arguments->filepath=arg;
			break;
		case 'c':
			arguments->flag=3;
			break;
		case 'l':
			arguments->flag=2;
			break;
		case 's':
			arguments->flag=1;
			break;
		case 'm':
			arguments->mounteddir=arg;
			break;
		case 't':
			arguments->deltime=arg;
			break;
		default:
			return ARGP_ERR_UNKNOWN;
	}
	return 0;
}

struct stat* getfilestat(char *path);


typedef struct expirer_data {
	int owner;
	unsigned long int inode_number,expiry_time;
	char *file_path;
	char *sha256sum;
} EXPIRER_DATA;
char md5sum[34];
char mounteddir[75];

int main(int argc,char *argv[]){
	ext2_filsys	current_fs = NULL;
	extern char device[75];
	char *filename;
	extern char mounteddir[75];
	int  open_flags = EXT2_FLAG_SOFTSUPP_FEATURES | EXT2_FLAG_RW;
	blk_t superblock=0;
	blk_t blocksize=0;
	int retval=0;
	int i=0,childpid,running;
	int ans=0,mins=60;
	struct arguments arguments;
	struct stat filestat;
	struct ext2_inode inode,*in={'\0'};
	struct ext2_inode_large *large_inode;
	extern char md5sum[34];
	EXPIRER_DATA user;

	/* Set defaults */
	arguments.device_name="";
	arguments.filepath="";
	arguments.flag=0;
	arguments.deltime="";
	arguments.mounteddir="";

	argp_parse (&argp, argc, argv, 0, 0, &arguments);

	/* Collect inputs */
	filename=arguments.filepath;
	if (arguments.deltime != "")
		mins=atoi(arguments.deltime);
	strcpy(device,arguments.device_name);

	if(arguments.flag == 2){
		expirer_list_files(0,filename);
		return 0;
	}
	if(arguments.flag == 3){
		if (filename == ""){
			printf("\n Usage : expirer -c -f absolutepath\n\n");
			return 1;
		}
		expirer_list_files(1,filename);
		return 0;
	}
	if(arguments.flag == 1){
		if (arguments.device_name == ""){
			printf("\n Usage : expirer -s -d devicename\n\n");
			return 1;
		}
		strcpy(mounteddir,arguments.mounteddir);
	}

	if (arguments.flag !=1 && stat(filename, &filestat) == -1)
		handle_error("file not found");
	// basic checks
	if (arguments.flag !=1 && getuid() != 0 && getuid() != filestat.st_uid ){
		printf("%d - %d",getuid(),filestat.st_uid);
		handle_error("Only owner can set expiry time\n");
	}

	if ( arguments.flag!=1 && (!S_ISREG(filestat.st_mode)))
		handle_error("Only regular file type supported.\n");


	retval = ext2fs_open(device, open_flags, superblock, blocksize,unix_io_manager, &current_fs);
	if (retval) {
		current_fs = NULL;
		handle_error("Error while opening filesystem.\n");
	}

	EXT2_BLOCK_SIZE=current_fs->blocksize;
	if (arguments.flag == 1){
		printf("\n Re-scanning drive %s for mountpoint%s\n",device,mounteddir);
		expirer_search_inodetype(current_fs,1); //regular files
		expirer_search_inodetype(current_fs,2); //directory inode

		//re-fill the data
		expirer_add_list_to_bdb(head);
		return 0;
	}
	expirer_read_inode(current_fs,&filestat,&inode);
	expirer_set_dtime(&inode,mins);
	expirer_write_inode(current_fs,&filestat,&inode);
	ext2fs_close(current_fs);
	//fill-in the values
	user.owner = filestat.st_uid;
	user.inode_number = filestat.st_ino ;
	user.expiry_time = inode.i_dtime;
	user.file_path = filename;
	expirer_compute_md5sum(filename);
	user.sha256sum = md5sum;

	expirer_bdb_insert(&user);
	
	running=system("pgrep expirerd &> /dev/null");
	if (running == 0){
	// If already running restart it via USR1 signal
	system("touch /var/run/expirerd.pid && kill -USR1 `cat /var/run/expirerd.pid` &> /dev/null");
	}else{
		system("/usr/sbin/expirerd &>/dev/null &");
	}

	fprintf(stdout,"\n File %s will expire in %d minutes : %s\n",user.file_path,mins,ctime(&user.expiry_time));
	return 0;
}
void expirer_add_list_to_bdb(struct node* current){
	EXPIRER_DATA user;
	struct stat filestat={'\0'};
	char *pathname=NULL;

	while(current!=NULL){

		user.inode_number = current->file_inode;
		user.expiry_time = current->expiry_time;
		user.file_path = expirer_get_abs_path(current->pathname);
		stat(user.file_path, &filestat);
		user.owner = filestat.st_uid;
		expirer_compute_md5sum(user.file_path);
		user.sha256sum = md5sum;

		expirer_bdb_insert(&user);
		current=current->next;
	}
}

int update_file_inode_list(struct node* current,unsigned long fileinode,unsigned long parent_inode,char *pathname){

	while(current!=NULL){
		if(fileinode == current->file_inode){
			current->parent_inode=parent_inode;
			current->pathname=pathname;
		}
		current=current->next;
	}
	return 1;
}
int search_file_inode_list(unsigned long fileinode){
	struct node* current=head;

	while(current!=NULL){
		if(fileinode == current->file_inode)
			return 0;
		current=current->next;
	}
	return 1;
}

struct private_data{
	unsigned long parent_inode;
	ext2_filsys current_fs;
};

char abs_path[75];
char* expirer_get_abs_path(char *path){
	extern char mounteddir[75],abs_path[75];
	memset(abs_path,0,75);
	strcpy(abs_path,mounteddir);
	return strcat(abs_path,path);
}
int expirer_check_entries(struct ext2_dir_entry *dirent,
			int offset EXT2FS_ATTR((unused)),
			int blocksize EXT2FS_ATTR((unused)),
			char *buf EXT2FS_ATTR((unused)), void *private)
{
	char *pathname=NULL;
	struct private_data *pnode = (struct private_data *)private;
	unsigned long file_inode= (unsigned long) dirent->inode;

	if(!search_file_inode_list(file_inode)){
		ext2fs_get_pathname (pnode->current_fs, pnode->parent_inode, dirent->inode, &pathname);
		update_file_inode_list(head,file_inode,pnode->parent_inode,pathname);
	}
}

/*
 * If inodetype = 1 , search only regular inodes and  create list of file inodes.
 * If inodetype = 2 , search only dir inodes and then lookup for file_inode# 
 */

int expirer_search_inodetype(ext2_filsys current_fs,int inodetype){
	ext2_inode_scan scan = 0;
	ext2_ino_t  ino;
	unsigned long count=0;
	struct ext2_inode inode,*in={'\0'};
	int retval=0,grp_desc=0;
	struct private_data *pino={'\0'};
	pino = calloc (1,sizeof(struct private_data));

	retval = ext2fs_open_inode_scan(current_fs, 0, &scan);
	if (retval) {
		handle_error("while opening inode scan");
	}


	do {
		retval = ext2fs_get_next_inode(scan, &ino, &inode);
	} while (retval == EXT2_ET_BAD_BLOCK_IN_INODE_TABLE);
	if (retval) {
		handle_error("error while starting inode scan");
	}

	while (ino){
		if (inodetype == 1){

			if (!LINUX_S_ISREG(inode.i_mode))
				goto next;

			if(inode.i_dtime && inode.i_links_count)
				expirer_file_inode_list_push(&head,ino,0,inode.i_dtime);

		}else{ 
			if (!LINUX_S_ISDIR(inode.i_mode))
				goto next;

			if (inode.i_dtime)
				goto next;
			pino->parent_inode=ino;
			pino->current_fs=current_fs;
			retval = ext2fs_dir_iterate(current_fs, ino, 0, 0,
					expirer_check_entries, pino);
		}

next:
		do {
			retval = ext2fs_get_next_inode(scan, &ino, &inode);
		} while (retval == EXT2_ET_BAD_BLOCK_IN_INODE_TABLE);
	}
	free(pino);
	if (scan)
		ext2fs_close_inode_scan(scan); 
}



//print the list
void expirer_printlist(struct node* current){
	while(current!=NULL){
		printf("\n file :%u",current->file_inode);
		printf("\n parent:%u",current->parent_inode);
		printf("\n dtime:%d",current->expiry_time);
		printf("\n pathname:%s",current->pathname);
		current=current->next;
	}
}

//push data into the list
void expirer_file_inode_list_push(struct node** headref,unsigned long file_inode,unsigned long parent_inode,unsigned long dtime){

	struct node* newnode=NULL;
	newnode=malloc(sizeof(struct node));

	newnode->file_inode=file_inode;
	newnode->parent_inode=parent_inode;
	newnode->expiry_time=dtime;

	newnode->next=*headref;

	*headref=newnode;
#if 0
	expirer_printlist(*headref);
#endif

}

void expirer_read_inode(ext2_filsys current_fs,struct stat* filestat,struct ext2_inode *in){
	ext2fs_read_inode(current_fs,filestat->st_ino,in);
}

void expirer_write_inode(ext2_filsys current_fs,struct stat* filestat,struct ext2_inode *in){
	int retval=0;
	retval=ext2fs_write_inode(current_fs,filestat->st_ino,in);
	if (retval) {
		current_fs = NULL;
		handle_error("Error while writing inode.");
	}
}

void expirer_set_dtime(struct ext2_inode *in,int minutes){
	if (minutes)
		in->i_dtime=(minutes*60)+expirer_current_time();
	else
		in->i_dtime=0;

}
/*
 * *expirer_list_files 
 * flag 0 - Just parse through the db using cursor.
 * flag 1 - Find given path and remove it from db.
 */
int expirer_list_files(int flag,char *filename){
	DBT key, data;
	DB *my_database;
	EXPIRER_DATA user;
	char *buffer,*key_buf;
	int ret;
	u_int32_t flags;
	DBC *cursorp;

	ret = db_create(&my_database, NULL, 0);
	if (ret != 0) {
		handle_error("db create fails.");
	}
	flags = DB_CREATE;   

	ret = my_database->open(my_database,NULL,DB_PATH,NULL,DB_BTREE,flags,0);
	if (ret != 0) {
		handle_error("db open fails.");
	}


	/* Zero out the DBTs before using them. */
	memset(&key, 0, sizeof(DBT));
	memset(&data, 0, sizeof(DBT));

	/* Get a cursor */
	my_database->cursor(my_database,NULL,&cursorp,0);

	if (!flag){
	printf("\n\tFilename\t      Expires on\n");
	printf("\n\t--------\t      ----------\n");
	}
	while ((ret = cursorp->get(cursorp,&key,&data,DB_NEXT)) == 0){
		key_buf = key.data;
		buffer = data.data;

		user.owner = *((int *)data.data);
		user.inode_number = *((unsigned long *)(buffer+sizeof(int)));
		user.expiry_time = *((unsigned long*)(buffer + sizeof(int) +sizeof(unsigned long int)));
		user.file_path = buffer + (sizeof(unsigned long int) * 2) + sizeof(int); 
		user.sha256sum = buffer + (sizeof(unsigned long int) * 2) + sizeof(int) + strlen(user.file_path) +1; 
		if (!flag){
			printf("\n\t%s\t",user.file_path);
			printf("%s",ctime(&user.expiry_time));
		}else{
			if (!strcmp(user.file_path,filename))
			{
			cursorp->del(cursorp, 0);
			if (cursorp != NULL)
			    cursorp->close(cursorp); 

			if (my_database != NULL)
			    my_database->close(my_database, 0);
			return 0;
			}
		}

	}
	if (cursorp != NULL)
	    cursorp->close(cursorp); 

	if (my_database != NULL)
	    my_database->close(my_database, 0);

	return 0;

}

void expirer_compute_md5sum(char *filelocation){
	FILE *pf;
	char md5_cmd[512];
	extern char md5sum[34];
	//Recompute md5  of recovered file
	memset(md5_cmd,'\0',512);
	memset(md5sum,'\0',34);
	sprintf(md5_cmd,"md5sum %s",filelocation);
	pf=popen(md5_cmd,"r");
	if(!pf){
		fprintf(stderr,"Could not open pipe");
		return ;
	}

	//get data
	fgets(md5sum, 34 , pf);

	if (pclose(pf) != 0)
		fprintf(stderr," Error: close Failed.");
}


// get current time in seconds.
time_t expirer_current_time(){
	struct timeval tv;
	struct timezone tz;

	gettimeofday(&tv, &tz);
	return tv.tv_sec;
}


int btreecompare(DB *db,const DBT *d1,const DBT *d2){
	unsigned long int d1_key = *(unsigned long int *)d1->data;
	unsigned long int d2_key = *(unsigned long int *)d2->data;

	if (d1_key < d2_key)
		return -1;
	else if (d1_key > d2_key)
		return 1;
	else	
		return 0;
}

int expirer_bdb_insert(EXPIRER_DATA *user){
	DBT key, data;
	DB *my_database;
	int buffsize, bufflen;
	char *databuff;
	u_int32_t flags;
	int ret;


	ret = db_create(&my_database, NULL, 0);
	if (ret != 0) {
		handle_error("db create failed");
	}

	/* Enable support for duplicate records */
	ret = my_database->set_flags(my_database,DB_DUPSORT);
	if (ret !=0 ){
		handle_error("Attempt to set duplication flag failed");
	}

	my_database->set_bt_compare(my_database,btreecompare);

	flags = DB_CREATE;   


	ret = my_database->open(my_database,NULL,DB_PATH,NULL,DB_BTREE,flags,0);
	if (ret != 0) {
		if (my_database != NULL)
			my_database->close(my_database, 0);
		handle_error("Open db failed");
	}


	/* Get the buffer */
	buffsize = sizeof(int) + sizeof(unsigned long)+ sizeof(unsigned long)+
		(strlen(user->file_path) + strlen(user->sha256sum) + 2);
	databuff = malloc(buffsize);
	memset(databuff, 0, buffsize);

	/* copy everything to the buffer */
	memcpy(databuff, &(user->owner), sizeof(int));
	bufflen = sizeof(int);

	memcpy(databuff + bufflen, &(user->inode_number),sizeof(unsigned long));
	bufflen += sizeof(unsigned long);

	memcpy(databuff + bufflen, &(user->expiry_time),sizeof(unsigned long));
	bufflen += sizeof(unsigned long);

	memcpy(databuff + bufflen, user->file_path, 
			strlen(user->file_path) + 1);
	bufflen += strlen(user->file_path) + 1;

	memcpy(databuff + bufflen, user->sha256sum, 
			strlen(user->sha256sum) + 1);
	bufflen += strlen(user->sha256sum) + 1;

	/* Zero out the DBTs before using them. */
	memset(&key, 0, sizeof(DBT));
	memset(&data, 0, sizeof(DBT));

	key.data = &(user->expiry_time);
	key.size = sizeof(unsigned long);

	data.data = databuff;
	data.size = bufflen;

	ret = my_database->put(my_database, NULL, &key, &data, 0);
	if (ret != 0 ){
		if (my_database != NULL)
			my_database->close(my_database, 0);
		handle_error("insert record failed");
	}

	/* When we're done with the database, close it. */
	if (my_database != NULL)
		my_database->close(my_database, 0);
	free(databuff);

}
